### PR TITLE
bugsnag: emit slice_name from service_name in custom metadata tab

### PR DESCRIPTION
**Pull request by morgan-wowk**
Some Bugsnag stacks filter on custom.slice_name to assign errors to
a slice. Store the service_name passed to setup() and write it as
custom.slice_name on every event so errors from orchestrator, api, and
metrics_poller each land in their own slice.

Consolidates all custom tab writes into a single add_tab call.

### CODE REVIEW

**Comment by morgan-wowk** [comment]
* **#268** <a href="https://app.graphite.com/github/pr/TangleML/tangle/268?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#267** <a href="https://app.graphite.com/github/pr/TangleML/tangle/267?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/TangleML/tangle/267?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#262** <a href="https://app.graphite.com/github/pr/TangleML/tangle/262?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#261** <a href="https://app.graphite.com/github/pr/TangleML/tangle/261?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: cloud_pipelines_backend/instrumentation/bugsnag_instrumentation.py [hunk 1]
```diff
@@ -102,8 +111,9 @@ def setup(*, service_name: str | None = None) -> None:
             project_root=service_name,
         )
         bugsnag_sdk.before_notify(_before_notify)
-        global _setup_called
+        global _setup_called, _service_name
         _setup_called = True
+        _service_name = service_name
```
Comment (Mbeaulne) [inline cloud_pipelines_backend/instrumentation/bugsnag_instrumentation.py:121]: > 🤖 This is an AI-generated code review comment.

**`setup()` docstring should mention the new `slice_name` side-effect.**

`service_name` used to be only the SDK's `project_root`. After this PR it is also stored as module state and surfaced on every event as `custom.slice_name`. Callers reading the existing docstring won't know that picking the `service_name` string also affects Bugsnag custom metadata on every event going forward.

**Fix:** one-line addition to the Args docstring:
```python
service_name: Identifies the process in Bugsnag (e.g. "tangle-api").
              Also surfaced as `slice_name` in the "custom" metadata tab on every event.
```

File: cloud_pipelines_backend/instrumentation/bugsnag_instrumentation.py [hunk 2]
```diff
@@ -50,7 +56,7 @@ def _before_notify(event: bugsnag_event.Event) -> None:
         )
         prefix = (event.metadata.get("extra") or {}).get("grouping_prefix")
         key_value = f"{prefix}: {chain}" if prefix else chain
-        event.add_tab("custom", {_CUSTOM_GROUPING_KEY: key_value})
+        custom[_CUSTOM_GROUPING_KEY] = key_value
```
Comment (Mbeaulne) [inline cloud_pipelines_backend/instrumentation/bugsnag_instrumentation.py:59]: > 🤖 This is an AI-generated code review comment.

**Silent key collision if operator sets `TANGLE_BUGSNAG_CUSTOM_GROUPING_KEY=slice_name`.**

```python
custom: dict[str, str] = {}
if _service_name:
    custom["slice_name"] = _service_name
...
custom[_CUSTOM_GROUPING_KEY] = key_value   # could overwrite "slice_name"
```

`_CUSTOM_GROUPING_KEY` is operator-supplied via env var. If someone sets it to `slice_name` — perhaps not realizing the new key exists — the grouping value will silently overwrite the slice name and the tab will only contain one of them.

**Fix:** low-effort options:
- Warn at module load: `if _CUSTOM_GROUPING_KEY == "slice_name": _logger.warning(...)` once, so the operator finds out.
- Or prefix Tangle-owned custom keys (`tangle_slice_name`) to make collisions almost impossible.
- Or just document the reserved key in the env-var comment block at the top of the file.

File: tests/instrumentation/test_bugsnag.py [hunk 3]
```diff
@@ -271,6 +271,31 @@ def test_before_notify_skips_error_class_prefix_gracefully_on_bad_errors_structu
...
+    with mock.patch("bugsnag.configure"), mock.patch("bugsnag.before_notify"):
+        bugsnag_module.setup(service_name="orchestrator")
+
+    mock_event = mock.MagicMock()
+    mock_event.original_error = None
+    mock_event.metadata = {}
+
+    bugsnag_module._before_notify(mock_event)
+
+    mock_event.add_tab.assert_called_once_with("custom", {"slice_name": "orchestrator"})
```
Comment (Mbeaulne) [inline tests/instrumentation/test_bugsnag.py:296]: > 🤖 This is an AI-generated code review comment.

**Missing negative-path test for `service_name=None` / no grouping key.**

The new code has two paths: (a) `_service_name` set → emit `slice_name`, (b) `_service_name` None AND no grouping key → skip the entire `add_tab("custom", ...)` call thanks to the `if custom:` guard at the bottom of `_before_notify`. Path (a) is tested by this test; path (b) isn't pinned anywhere. A future refactor could silently start emitting an empty `custom` tab on every event without any test failing.

**Fix:** add a sibling test:
```python
def test_before_notify_omits_custom_tab_when_no_service_name_and_no_grouping_key(monkeypatch):
    monkeypatch.setenv("TANGLE_BUGSNAG_API_KEY", "test-api-key")
    monkeypatch.delenv("TANGLE_BUGSNAG_CUSTOM_GROUPING_KEY", raising=False)
    importlib.reload(bugsnag_module)
    contextual_logging.clear_context_metadata()
    with mock.patch("bugsnag.configure"), mock.patch("bugsnag.before_notify"):
        bugsnag_module.setup(service_name=None)

    mock_event = mock.MagicMock()
    mock_event.original_error = None
    mock_event.metadata = {}
    bugsnag_module._before_notify(mock_event)

    custom_calls = [c for c in mock_event.add_tab.call_args_list if c.args[0] == "custom"]
    assert custom_calls == []
```

**Review by yuechao-qin — APPROVED** [review]

File: cloud_pipelines_backend/instrumentation/bugsnag_instrumentation.py (hunk 1 above)
Comment (morgan-wowk) [reply cloud_pipelines_backend/instrumentation/bugsnag_instrumentation.py:121]: 🤖 Updated the `setup()` docstring to document both uses of `service_name`: the existing Bugsnag `project_root` identification and the new `slice_name` custom metadata emission. Also added a note reserving the `slice_name` key to prevent silent collision with `TANGLE_BUGSNAG_CUSTOM_GROUPING_KEY`.

File: cloud_pipelines_backend/instrumentation/bugsnag_instrumentation.py (hunk 2 above)
Comment (morgan-wowk) [reply cloud_pipelines_backend/instrumentation/bugsnag_instrumentation.py:59]: 🤖 Addressed via the docstring — added an explicit note that `slice_name` is a reserved key and operators should not set `TANGLE_BUGSNAG_CUSTOM_GROUPING_KEY=slice_name`. A runtime warning felt like over-engineering for a misconfiguration that would only affect one operator on one deploy; the doc note should catch it at setup time.

File: tests/instrumentation/test_bugsnag.py (hunk 3 above)
Comment (morgan-wowk) [reply tests/instrumentation/test_bugsnag.py:296]: 🤖 Added `test_before_notify_omits_custom_tab_when_no_service_name_and_no_grouping_key` — sets up with `service_name=None` and no `TANGLE_BUGSNAG_CUSTOM_GROUPING_KEY`, then asserts no `custom` tab is emitted. Pins the `if custom:` guard against future regressions.

**Review by Mbeaulne — APPROVED** [review]

**Comment by morgan-wowk** [comment]
### Merge activity

* **Jun 13, 12:04 AM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/TangleML/tangle/267).
* **Jun 13, 12:08 AM UTC**: [Graphite](https://app.graphite.com/github/pr/TangleML/tangle/267) rebased this pull request as part of a merge.
* **Jun 13, 12:08 AM UTC**: @morgan-wowk merged this pull request with [Graphite](https://app.graphite.com/github/pr/TangleML/tangle/267).
